source/common/router/router.cc

```diff
@@ -59,12 +59,57 @@ bool FilterUtility::shouldShadow(const ShadowPolicy& policy, Runtime::Loader& ru
 
 FilterUtility::TimeoutData
 FilterUtility::finalTimeout(const RouteEntry& route, Http::HeaderMap& request_headers,
-                            bool insert_envoy_expected_request_timeout_ms) {
-  // See if there is a user supplied timeout in a request header. If there is we take that,
-  // otherwise we use the default.
+                            bool insert_envoy_expected_request_timeout_ms, bool grpc_request) {
+  // See if there is a user supplied timeout in a request header. If there is we take that.
+  // Otherwise we check if the request is gRPC and use the timeout in the gRPC headers or
+  // infinity when gRPC headers have no timeout, or the default from the route config for
+  // non-gRPC requests.
   TimeoutData timeout;
-  timeout.global_timeout_ = route.timeout();
-  timeout.per_try_timeout_ = route.retryPolicy().perTryTimeout();
+  if (grpc_request && route.useGrpcTimeout()) {
+    timeout.global_timeout_ = std::chrono::milliseconds(0);
+    timeout.per_try_timeout_ = std::chrono::milliseconds(0);
+
+    Http::HeaderEntry* header_grpc_timeout_entry = request_headers.GrpcTimeout();
+    if (header_grpc_timeout_entry) {
+      uint64_t header_grpc_timeout;
```

Comment:
extract this part to a function parseGrpcTimeout

Reply:
Added `std::chrono::milliseconds Common::getGrpcTimeout(Http::HeaderMap& request_headers)`